runtime/router.go

```diff
@@ -108,103 +44,70 @@ type HandlerFn func(
 	*ServerHTTPResponse,
 )
 
-// EndpointMetrics contains pre allocated metrics structures
-// These are pre-allocated to cache tags maps and for performance
-type EndpointMetrics struct {
-	requestRecvd   tally.Counter
-	requestLatency tally.Timer
-	statusCodes    map[int]tally.Counter
-}
-
 // RouterEndpoint struct represents an endpoint that can be registered
 // into the router itself.
 type RouterEndpoint struct {
-	EndpointName string
-	HandlerName  string
-	HandlerFn    HandlerFn
-
-	metrics EndpointMetrics
-	gateway *Gateway
+	HandlerFn HandlerFn
+	metrics   *InboundHTTPMetrics
+	gateway   *Gateway
 }
 
 // NewRouterEndpoint creates an endpoint with all the necessary data
 func NewRouterEndpoint(
 	gateway *Gateway,
-	endpointName string,
-	handlerName string,
 	handler HandlerFn,
+	scope tally.Scope,
 ) *RouterEndpoint {
-	endpointTags := map[string]string{
-		"endpoint": endpointName,
-		"handler":  handlerName,
-	}
-	endpointScope := gateway.AllHostScope.Tagged(endpointTags)
-	requestRecvd := endpointScope.Counter("inbound.calls.recvd")
-	requestLatency := endpointScope.Timer("inbound.calls.latency")
-	statusCodes := make(map[int]tally.Counter, len(knownStatusCodes))
-
-	for _, statusCode := range knownStatusCodes {
-		metricName := "inbound.calls.status." + strconv.Itoa(statusCode)
-		statusCodes[statusCode] = endpointScope.Counter(metricName)
-	}
-
 	return &RouterEndpoint{
-		EndpointName: endpointName,
-		HandlerName:  handlerName,
-		HandlerFn:    handler,
-		gateway:      gateway,
-
-		metrics: EndpointMetrics{
-			requestRecvd:   requestRecvd,
-			statusCodes:    statusCodes,
-			requestLatency: requestLatency,
-		},
+		HandlerFn: handler,
+		gateway:   gateway,
+		metrics:   NewInboundHTTPMetrics(scope, gateway.Logger),
 	}
 }
 
 // HandleRequest is called by the router and starts the request
 func (endpoint *RouterEndpoint) HandleRequest(
 	w http.ResponseWriter, r *http.Request, params httprouter.Params,
 ) {
-	reqFields := logRequestFields(r)
-	var resFields []zapcore.Field
-
+	var req *ServerHTTPRequest
 	defer func() {
-		writeLogs(endpoint.gateway.Logger, reqFields, resFields)
+		// Post-process logs & metrics
+		writeLogs(endpoint.gateway.Logger, logRequestFields(r), logResponseFields(req.Res))
+		endpoint.metrics.CollectMetrics(req.StartTime, req.Res.FinishTime, req.Res.StatusCode)
```

Comment:
Deferring the collection of metrics loses information.

We really want to emit the metrics in the right place at the start and finish methods of the request and response class.

Reply:
We don't loose the metrics in the defer.  Defer is also run at the end of the function call.